pkg/container/queue/chunkqueue.go

```diff
@@ -0,0 +1,389 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package queue
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/pingcap/log"
+)
+
+const (
+	// the size of each chunk is 1024 bytes (1kB) by default
+	defaultSizePerChunk = 1024
+	// the minimum length of each chunk is 16
+	minimumChunkLen      = 16
+	defaultPitchArrayLen = 16
+)
+
+// ChunkQueue is a generic, efficient, iterable and GC-friendly queue.
+// Attention, it's not thread-safe.
+type ChunkQueue[T any] struct {
+	// [head, tail) is the section of chunks in use
+	head int
+	tail int
+
+	// size is number of elements in queue
+	size int
+
+	// chunks is an array storing ptr
+	chunks []*chunk[T]
+	// chunkLength is the max number of elements stored in every chunk
+	chunkLength  int
```

Comment:
It seems that the `chunkLength` might be confusing, can we use the `chunkMaxLenght`? Because it reminds me it is the length of chunks.

Reply:
The length of a chunk is a constant for a queue. I think using chunkMaxLength implies it's changeable and the size of memory allocation per chunk varies. 
In fact, it's a private variable since any details of chunks are oblivious to programmers who use ChunkQueue.